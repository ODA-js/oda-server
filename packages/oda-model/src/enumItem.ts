import {
  ModelBase,
  ModelBaseInput,
  ModelBaseInternal,
  IModelBase,
  ModelBaseOutput,
  ModelBaseMetaInfo,
} from './modelbase';
import { merge } from 'lodash';
import { MetaModelType, Nullable, assignValue } from './types';
import { Internal } from './element';

export interface IEnumItem
  extends IModelBase<EnumItemMetaInfo, EnumItemInput, EnumItemOutput> {
  readonly value: string;
}

export interface EnumItemMetaInfo extends ModelBaseMetaInfo {}

export interface EnumItemInternal extends ModelBaseInternal {
  value: string;
}

export interface EnumItemInput extends ModelBaseInput<EnumItemMetaInfo> {
  value?: string;
}
export interface EnumItemOutput extends ModelBaseOutput<EnumItemMetaInfo> {
  value?: string;
}

export const enumItemDefaultMetaInfo = {};
export const enumItemDefaultInput = { metadata: enumItemDefaultMetaInfo };

export class EnumItem
  extends ModelBase<
    EnumItemMetaInfo,
    EnumItemInput,
    EnumItemInternal,
    EnumItemOutput
  >
  implements IEnumItem {
  public get modelType(): MetaModelType {
    return 'enum-item';
  }

  constructor(init: EnumItemInput) {
    super(merge({}, enumItemDefaultInput, init));
  }

  get value(): string {
    return this[Internal].value || this[Internal].name;
  }

  updateWith(input: Nullable<EnumItemInput>) {
    super.updateWith(input);
    assignValue<EnumItemInternal, EnumItemInput, EnumItemInput['value']>({
      src: this[Internal],
      input,
      field: 'value',
    });
  }

  toObject(): EnumItemOutput {
    return merge({}, super.toObject(), {
      value: this[Internal].value,
    } as Partial<EnumItemOutput>);
  }
  mergeWith(payload: Nullable<EnumItemInput>) {
    super.mergeWith(payload);
  }
}
