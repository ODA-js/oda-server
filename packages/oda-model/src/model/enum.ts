import clean from '../lib/json/clean';
import { MetaModelType, EnumStorage, EnumInput } from './interfaces';
import { ModelBase } from './modelbase';
import { IEnum, IEnumItemsInput } from './interfaces/model';

export class Enum extends ModelBase implements IEnum {
  public modelType: MetaModelType = 'union';
  protected $obj!: EnumStorage;

  get items(): IEnumItemsInput[] {
    return this.$obj.items;
  }

  public updateWith(obj: EnumInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };
      result.name = obj.name;
      let $items = obj.items;
      let items = $items;

      result.items = items
        .map(i => (typeof i === 'string' ? { name: i } : i))
        .map(i => ({
          name: i.name,
          title: i.title || i.name,
          description: i.description || i.title || i.name,
          value: i.value,
          metadata: i.metadata,
        }));

      result.items_ = $items;

      this.$obj = result;
    }
  }

  // it get fixed object
  public toObject() {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      items: props.items,
    });
  }
}
