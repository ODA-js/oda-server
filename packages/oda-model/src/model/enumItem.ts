import {
  MetaModelType,
  EnumItemMeta,
  EnumItemInput,
  EnumItemStorage,
  IEnumItem,
} from './interfaces';
import { ModelBase } from './modelbase';

export class EnumItem
  extends ModelBase<EnumItemMeta, EnumItemInput, EnumItemStorage>
  implements IEnumItem {
  public modelType: MetaModelType = 'enum-item';
  constructor(inp: EnumItemInput) {
    super(inp);
  }
  get value(): string {
    return this.$obj.value;
  }
  updateWith(inp: EnumItemInput) {
    super.updateWith(inp);
    this.$obj.value = inp.value || '';
  }
  toObject(): EnumItemInput {
    return {
      ...super.toObject(),
      value: this.value,
    };
  }
}
