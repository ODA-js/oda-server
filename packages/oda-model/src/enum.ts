import {
  ModelBase,
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBaseOutput,
} from './modelbase';
import { EnumItem, IEnumItem, EnumItemInput } from './enumItem';
import { merge } from 'lodash';
import { ElementMetaInfo } from './element';
import { MetaModelType, Nullable, assignValue } from './types';

export interface IEnum extends IModelBase<EnumMetaInfo, EnumInput, EnumOutput> {
  /**
   * Enum item definition
   */
  readonly items: Map<string, IEnumItem>;
}

export interface EnumMetaInfo extends ElementMetaInfo {}

export interface EnumInternal extends ModelBaseInternal<EnumMetaInfo> {
  items: Map<string, IEnumItem>;
}

export interface EnumInput extends ModelBaseInput<EnumMetaInfo> {
  items: (EnumItemInput | string)[];
}

export interface EnumOutput extends ModelBaseOutput<EnumMetaInfo> {
  items: EnumItemInput[];
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Enum
  extends ModelBase<EnumMetaInfo, EnumInput, EnumInternal, EnumOutput>
  implements IEnum {
  public get modelType(): MetaModelType {
    return 'enum';
  }
  get items() {
    return this.$obj.items;
  }

  constructor(init: EnumInput) {
    super(merge({}, defaultInput, init));
  }

  public updateWith(input: Nullable<EnumInput>) {
    super.updateWith(input);

    assignValue<EnumInternal, EnumInput, EnumInput['items']>({
      src: this.$obj,
      input,
      field: 'items',
      effect: (src, value) =>
        (src.items = new Map(
          value
            .map(i =>
              typeof i === 'string' ? ({ name: i } as EnumItemInput) : i,
            )
            .map(
              i =>
                [
                  i.name,
                  new EnumItem({
                    name: i.name,
                    title: i.title || i.name,
                    description: i.description || i.title || i.name,
                    value: i.value,
                    metadata: i.metadata,
                  }),
                ] as [string, IEnumItem],
            ),
        )),
      required: true,
    });
  }

  public toObject(): EnumOutput {
    return merge({}, super.toObject(), {
      items: [...this.$obj.items.values()].map(i => i.toObject()),
    } as Partial<EnumOutput>);
  }
}
