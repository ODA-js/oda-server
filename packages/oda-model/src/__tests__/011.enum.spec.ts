import 'jest';
import { Enum } from '../enum';
import { enumItemDefaultMetaInfo } from '../enumItem';

describe('Enum', () => {
  it('default', () => {
    const res = new Enum({
      name: 'LOCATION',
      items: [{ name: 'HOME' }, { name: 'WORk' }],
    });
    expect(res.metadata).toMatchObject(enumItemDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('dedupes items when it is strings', () => {
    const res = new Enum({
      name: 'location',
      items: ['WORK', 'WORK', 'HOME'],
    });
    expect(res.items.size).toBe(2);
  });
  it('dedupes items when it is objects and string', () => {
    const res = new Enum({
      name: 'location',
      items: ['WORK', { name: 'WORK' }, 'HOME'],
    });
    expect(res.items.size).toBe(2);
  });
  it('dedupes items when it is objects ', () => {
    const res = new Enum({
      name: 'location',
      items: [
        { name: 'WORK', description: 'work place' },
        { name: 'WORK' },
        { name: 'WORK', value: '_WORK_' },
        'HOME',
      ],
    });
    expect(res.items.size).toBe(2);
    const item = res.items.get('WORK');
    expect(item).not.toBeUndefined();
    if (item) {
      expect(item.description).toBe('work place');
      expect(item.value).toBe('_WORK_');
    }
  });
  it('merges', () => {
    const res1 = new Enum({
      name: 'location1',
      items: ['WORK1', 'WORK', 'HOME'],
    });
    const res2 = new Enum({
      name: 'localtion2',
      items: [
        'WORK2',
        { name: 'WORK', description: 'just work location' },
        'HOME',
      ],
    });
    const res = new Enum({
      name: 'location',
      items: ['WORK', 'WORK', 'HOME'],
    });
    expect(res.name).toBe('location');
    res.mergeWith(res1.toObject());
    res.mergeWith(res2.toObject());
    expect(res.items.size).toBe(4);
    expect(res.toObject()).toMatchSnapshot('merge');
  });
});
