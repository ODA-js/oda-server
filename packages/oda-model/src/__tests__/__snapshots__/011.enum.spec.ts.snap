// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enum default: toObject 1`] = `
Object {
  "description": undefined,
  "items": Array [
    Object {
      "description": undefined,
      "metadata": Object {},
      "name": "HOME",
      "title": undefined,
      "value": undefined,
    },
    Object {
      "description": undefined,
      "metadata": Object {},
      "name": "WORk",
      "title": undefined,
      "value": undefined,
    },
  ],
  "metadata": Object {},
  "name": "LOCATION",
  "title": undefined,
}
`;

exports[`Enum merges: merge 1`] = `
Object {
  "description": undefined,
  "items": Array [
    Object {
      "description": "just work location",
      "metadata": Object {},
      "name": "WORK",
      "title": undefined,
      "value": undefined,
    },
    Object {
      "description": undefined,
      "metadata": Object {},
      "name": "HOME",
      "title": undefined,
      "value": undefined,
    },
    Object {
      "description": undefined,
      "metadata": Object {},
      "name": "WORK1",
      "title": undefined,
      "value": undefined,
    },
    Object {
      "description": undefined,
      "metadata": Object {},
      "name": "WORK2",
      "title": undefined,
      "value": undefined,
    },
  ],
  "metadata": Object {},
  "name": "location",
  "title": undefined,
}
`;
